src/adjustproxyconnection.cpp

```diff
@@ -0,0 +1,196 @@
+    : QObject(parent), m_connection(connection) {
+  MVPN_COUNT_CTOR(AdjustProxyConnection);
+
+  logger.debug() << "New connection received";
+
+  Q_ASSERT(m_connection);
+  connect(m_connection, &QTcpSocket::readyRead, this,
+          &AdjustProxyConnection::readData);
+}
+
+AdjustProxyConnection::~AdjustProxyConnection() {
+  MVPN_COUNT_DTOR(AdjustProxyConnection);
+  logger.debug() << "Connection released";
+}
+
+void AdjustProxyConnection::readData() {
+  Q_ASSERT(m_connection);
+  QByteArray input = m_connection->readAll();
+  m_buffer.append(input);
+
+  switch (m_state) {
+    case ProcessingState::NotStarted:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processFirstLine();
+      [[fallthrough]];
+    case ProcessingState::FirstLineDone:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processHeaders();
+      [[fallthrough]];
+    case ProcessingState::HeadersDone:
+      processParameters();
+      [[fallthrough]];
+    case ProcessingState::ParametersDone:
+      filterParametersAndForwardRequest();
+      break;
+    default:
+      Q_ASSERT(false);
+      logger.debug() << "Unknown ProcessingState: " << m_state;
+  }
+}
+
+void AdjustProxyConnection::processFirstLine() {
+  int pos = m_buffer.indexOf("\n");
+  if (pos == -1) {
+    return;
+  }
+
+  QByteArray line = m_buffer.left(pos);
+  m_buffer.remove(0, pos + 1);
+
+  QList<QByteArray> parts = line.split(' ');
+  if (parts.length() < 2) {
+    logger.info() << "Invalid HTTP request";
+    m_connection->close();
+    return;
+  }
+
+  m_method.append(parts[0]);
+  m_route = parts[1].trimmed();
+
+  m_state = ProcessingState::FirstLineDone;
+}
+
+void AdjustProxyConnection::processHeaders() {
+  while (true) {
+    int pos = m_buffer.indexOf("\n");
+    if (pos == -1) {
+      break;
+    }
+
+    QByteArray line = m_buffer.left(pos);
+    m_buffer.remove(0, pos + 1);
+
+    QByteArray header = line.trimmed();
+    if (header.isEmpty()) {
+      break;
+    }
+
+    pos = header.indexOf(":");
+    if (pos == -1) {
+      continue;
+    }
+
+    QByteArray headerName = header.left(pos);
+    header.remove(0, pos + 1);
+    QByteArray headerValue = header.trimmed();
+
+    QPair<QString, QString> headerPair;
+    headerPair.first = QString(headerName);
+    headerPair.second = QString(headerValue);
+
+    m_headers.append(headerPair);
+  }
+
+  m_state = ProcessingState::HeadersDone;
+}
+
+void AdjustProxyConnection::processParameters() {
+  if (m_method == "GET") {
+    m_parameters = QUrlQuery(m_route);
+  } else {
+    if (m_buffer.isEmpty()) {
+      return;
+    }
+    m_parameters = QUrlQuery(m_buffer.trimmed());
+  }
+
+  m_state = ProcessingState::ParametersDone;
+}
+
+void AdjustProxyConnection::filterParametersAndForwardRequest() {
+  QList<QPair<QString, QString>> newParameters;
+
+  for (QPair<QString, QString> parameter : m_parameters.queryItems()) {
+    if (parameter.first == "adid" || parameter.first == "app_token" ||
```

Comment:
Here, I would suggest something different. Let's create 2 QList<> of params (allowed, denied). Anything else should be kept in a list, written in the log file, and shared with the guardian-website.